### PR TITLE
update rules to remove segment in leader view

### PR DESCRIPTION
**Pull request by weiliu1031**
Signed-off-by: Wei Liu <wei.liu@zilliz.com>
issue: #22874

### CODE REVIEW

File: internal/querycoordv2/observers/leader_observer.go [hunk 1]
```diff
@@ -168,18 +168,15 @@ func (o *LeaderObserver) findNeedRemovedSegments(leaderView *meta.LeaderView, di
...
-		if ok || existInCurrentTarget || existInNextTarget {
+		if ok {
 			continue
 		}
 		log.Debug("leader observer append a segment to remove:", zap.Int64("collectionID", leaderView.CollectionID),
 			zap.String("Channel", leaderView.Channel), zap.Int64("leaderViewID", leaderView.ID),
-			zap.Int64("segmentID", sid), zap.Bool("distMap_exist", ok),
-			zap.Bool("existInCurrentTarget", existInCurrentTarget),
-			zap.Bool("existInNextTarget", existInNextTarget))
+			zap.Int64("segmentID", sid), zap.Bool("distMap_exist", ok))
```
Comment (yah01) [inline internal/querycoordv2/observers/leader_observer.go:179]: Please place the fields into lines by each one

Comment (yah01) [reply internal/querycoordv2/observers/leader_observer.go:179]: Also the field "distMap_exist" is not needed, as it's always false here

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/querycoordv2/observers/leader_observer.go [hunk 2]
```diff
@@ -168,18 +168,15 @@ func (o *LeaderObserver) findNeedRemovedSegments(leaderView *meta.LeaderView, di
...
+	// if a segment(segment_id, node_id) only exists in the leader view, it must be cleaned up
 	for sid, s := range leaderView.Segments {
 		_, ok := distMap[sid]
-		existInCurrentTarget := o.target.GetHistoricalSegment(leaderView.CollectionID, sid, meta.CurrentTarget) != nil
-		existInNextTarget := o.target.GetHistoricalSegment(leaderView.CollectionID, sid, meta.NextTarget) != nil
-		if ok || existInCurrentTarget || existInNextTarget {
+		if ok {
 			continue
 		}
 		log.Debug("leader observer append a segment to remove:", zap.Int64("collectionID", leaderView.CollectionID),
```
Comment (yah01) [inline internal/querycoordv2/observers/leader_observer.go:177]: the log could be simplifed: "stale leader view segment found", the ":" symbol is not needed

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/querycoordv2/observers/leader_observer.go [hunk 3]
```diff
@@ -168,18 +168,17 @@ func (o *LeaderObserver) findNeedRemovedSegments(leaderView *meta.LeaderView, di
...
 			continue
 		}
-		log.Debug("leader observer append a segment to remove:", zap.Int64("collectionID", leaderView.CollectionID),
-			zap.String("Channel", leaderView.Channel), zap.Int64("leaderViewID", leaderView.ID),
-			zap.Int64("segmentID", sid), zap.Bool("distMap_exist", ok),
-			zap.Bool("existInCurrentTarget", existInCurrentTarget),
-			zap.Bool("existInNextTarget", existInNextTarget))
+		log.Debug("stale leader view segment found",
+			zap.Int64("collectionID", leaderView.CollectionID),
+			zap.String("Channel", leaderView.Channel),
```
Comment (yah01) [inline internal/querycoordv2/observers/leader_observer.go:179]: should start with lowercase letter, -> "channel"

File: internal/querycoordv2/observers/leader_observer.go [hunk 4]
```diff
@@ -168,18 +168,17 @@ func (o *LeaderObserver) findNeedRemovedSegments(leaderView *meta.LeaderView, di
...
+			zap.Int64("leaderViewID", leaderView.ID),
```
Comment (yah01) [inline internal/querycoordv2/observers/leader_observer.go:180]: leaderID is fine, as it's actually the node ID of leader

**Review by yah01 — CHANGES_REQUESTED** [review]

**Review by congqixia — COMMENTED** [review]
We cannot remove the leader view if the segment is still in target. Maybe adding some extra info when fetching distribution to mark leader offline is more reasonable?
/hold

**Pull request closed without merging**